gradio/chat_interface.py

```diff
@@ -258,6 +258,7 @@ def __init__(
             self.chatbot_state = (
                 State(self.chatbot.value) if self.chatbot.value else State([])
             )
+            self.previous_input = State(value=[])
```

Comment:
The cause of #9227 is that hitting `undo` will update `saved_input` to be the undone user message. But then that will be used as the input of `retry` which is not what we want. 

We want `saved_input` to be the latest submitted user message. In order to keep track of that, I'm adding `previous_input` variable.